providers/implementations/encode_decode/ml_kem_codecs.c

```diff
@@ -0,0 +1,559 @@
+    }
+    publen = key->vinfo->pubkey_bytes;
+
+    if (out != NULL
+        && (*out = OPENSSL_malloc(publen)) == NULL)
+        return 0;
+    if (!ossl_ml_kem_encode_public_key(*out, publen, key)) {
+        ERR_raise_data(ERR_LIB_OSSL_ENCODER, ERR_R_INTERNAL_ERROR,
+                       "error encoding %s public key",
+                       key->vinfo->algorithm_name);
+        OPENSSL_free(*out);
+        return 0;
+    }
+
+    return (int)publen;
+}
+
+/* Allocate and encode PKCS#8 private key payload. */
+int ossl_ml_kem_i2d_prvkey(const ML_KEM_KEY *key, uint8_t **out,
+                           PROV_CTX *provctx)
+{
+    const ML_KEM_VINFO *v = key->vinfo;
+    const ML_KEM_CINFO *c;
+    ML_KEM_PKCS8_PREF *vp8_alloc, *vp8_slot;
+    const ML_KEM_PKCS8_INFO *vp8;
+    uint8_t *buf = NULL, *pos;
+    const char *formats;
+    int len = ML_KEM_SEED_BYTES;
+    int ret = 0;
+
+    /* Not ours to handle */
+    if ((c = ml_kem_get_cinfo(v->evp_type)) == NULL)
+        return 0;
+
+    if (!ossl_ml_kem_have_prvkey(key)) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_NOT_A_PRIVATE_KEY,
+                       "no %s private key data available",
+                       key->vinfo->algorithm_name);
+        return 0;
+    }
+
+    formats = ossl_prov_ctx_get_param(
+        provctx, OSSL_PKEY_PARAM_ML_KEM_OUTPUT_FORMATS, NULL);
+    vp8_slot = vp8_alloc = vp8_order(v->algorithm_name, c->pkcs8_info,
+                                     "output", formats);
+    if (vp8_alloc == NULL)
+        return 0;
+
+    /* If we don't have a seed, skip seedful entries */
+    if (!ossl_ml_kem_have_seed(key))
+        while (vp8_slot->vp8_entry != NULL
+               && vp8_slot->vp8_entry->seed_offset != 0)
+            ++vp8_slot;
+    /* No matching table entries, give up */
+    if ((vp8 = vp8_slot->vp8_entry) == NULL) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_ML_KEM_NO_FORMAT,
+                       "no matching enabled %s private key output formats",
+                       v->algorithm_name);
+        goto end;
+    }
+    len = vp8->p8_bytes;
+
+    if (out == NULL) {
+        ret = len;
+        goto end;
+    }
+
+    if ((pos = buf = OPENSSL_malloc((size_t) len)) == NULL)
+        goto end;
+
+    pos = OPENSSL_store_u32_be(pos, vp8->p8_magic);
+    if (vp8->seed_offset != 0) {
+        /*
+         * Either the tag/len were already included in |magic| or they require
+         * us to write two bytes now.
+         */
+        if (pos == buf + vp8->seed_offset - 2)
+            pos = OPENSSL_store_u16_be(pos, vp8->seed_magic);
+        if (pos != buf + vp8->seed_offset
+            || !ossl_ml_kem_encode_seed(pos, ML_KEM_SEED_BYTES, key)) {
+            ERR_raise_data(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR,
+                           "error encoding %s private key",
+                           v->algorithm_name);
+            goto end;
+        }
+        pos += ML_KEM_SEED_BYTES;
+    }
+    if (vp8->priv_offset != 0) {
```

Comment:
My opinion is that this is the wrong thing to do on output.
Loading someone else's crappy format is acceptable for interop,
but I think this is taking it too far here by also writing it..
We should be just saving out openssl format here.

Reply:
This may also mean there is no need for sorting.

Reply:
The sorting is a way to arrange the table entries we support to be a the front of the list, so the others are easily skipped.
It is useful even if all it does is support truncation of the list, with the order of the remaining entries not significant.
Why are we hung up on sorting wanted entries before unwanted entries?